python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,292 @@
+        self.stats = stats
+        return self
+
+    @property
+    def type(self):
+        cdef CFileType ctype = self.stats.type()
+        if ctype == CFileType_NonExistent:
+            return FileType.NonExistent
+        elif ctype == CFileType_Unknown:
+            return FileType.Uknown
+        elif ctype == CFileType_File:
+            return FileType.File
+        elif ctype == CFileType_Directory:
+            return FileType.Directory
+        else:
+            raise ValueError('Unhandled FileType {}'.format(type))
+
+    @property
+    def path(self):
+        return Path(frombytes(self.stats.path()))
+
+    @property
+    def base_name(self):
+        return frombytes(self.stats.base_name())
+
+    @property
+    def size(self):
+        return self.stats.size()
+
+    @property
+    def extension(self):
+        return frombytes(self.stats.extension())
+
+    # @property
+    # def mtime(self):
+    #     return self.stats.mtime()
+
+
+cdef class Selector:
+
+    cdef CSelector selector
+
+    def __init__(self, base_dir='', bint allow_non_existent=False,
+                 bint recursive=False):
+        self.base_dir = base_dir
+        self.recursive = recursive
+        self.allow_non_existent = allow_non_existent
+
+    @property
+    def base_dir(self):
+        return Path(self.selector.base_dir)
+
+    @base_dir.setter
+    def base_dir(self, base_dir):
+        self.selector.base_dir = _to_string(base_dir)
+
+    @property
+    def allow_non_existent(self):
+        return self.selector.allow_non_existent
+
+    @allow_non_existent.setter
+    def allow_non_existent(self, bint allow_non_existent):
+        self.selector.allow_non_existent = allow_non_existent
+
+    @property
+    def recursive(self):
+        return self.selector.recursive
+
+    @recursive.setter
+    def recursive(self, bint recursive):
+        self.selector.recursive = recursive
+
+
+cdef class FileSystem:
+
+    cdef:
+        shared_ptr[CFileSystem] wrapped
+        CFileSystem* fs
+
+    def __init__(self):
+        raise TypeError('dont initialize me')
+
+    cdef init(self, const shared_ptr[CFileSystem]& wrapped):
+        self.wrapped = wrapped
+        self.fs = wrapped.get()
+
+    def stat(self, paths_or_selector):
+        cdef:
+            vector[CFileStats] stats
+            vector[c_string] paths
+            CSelector selector
+
+        if isinstance(paths_or_selector, Selector):
+            selector = (<Selector>paths_or_selector).selector
+            check_status(self.fs.GetTargetStats(selector, &stats))
+        elif isinstance(paths_or_selector, (list, tuple)):
+            paths = [_to_string(s) for s in paths_or_selector]
+            check_status(self.fs.GetTargetStats(paths, &stats))
+        else:
+            raise TypeError('Must pass either paths or a Selector')
+
+        return [FileStats.wrap(stat) for stat in stats]
+
+    def mkdir(self, path, bint recursive=True):
```

Comment:
I would favor reusing the C++ names (but converted to PEP8).

Reply:
+1